fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+
+def run_fuzzer(branch, t, gcs, exit_event):
+    # First, figure out the current corpus version
+    bucket = gcs.get_bucket(GCS_BUCKET)
+    corpus_vers = bucket.blob('current-corpus').download_as_text().strip()
+
+    date = datetime.datetime.now().strftime('%Y-%m-%d')
+    log_path = pathlib.Path(date) / t['crate'] / t['runner'] / str(uuid.uuid4())
+    log_filename = WORKDIR / 'fuzz-logs' / log_path
+    utils.mkdirs(log_filename.parent)
+    with open(log_filename, 'a') as log_file:
+        current_commit = subprocess.run(['git', 'rev-parse', 'HEAD'], cwd=REPO_DIR, capture_output=True, check=True).stdout
+        log_file.write(f"Corpus version: {corpus_vers}\n")
+        log_file.write(f"Running on commit {current_commit}\n")
+        log_file.flush()
+
+        # Then, make sure the local corpus is the same as the gcs version
+        reset_to_gcs(t['crate'] + '/corpus/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+        reset_to_gcs(t['crate'] + '/artifacts/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+
+        # Then, make sure we auto-upload any new discovery to gcs
+        stop_inotify_event = threading.Event()
+        auto_upload_to_gcs(
+            stop_inotify_event,
+            t['crate'] + '/corpus/' + t['runner'],
+            REPO_DIR,
+            corpus_vers,
+            bucket,
+            log_file,
+            False,
+            lambda path: FUZZ_CORPUS_UPLOADED.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc(),
+            lambda path: FUZZ_CORPUS_DELETED.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc(),
+        )
+        auto_upload_to_gcs(
+            stop_inotify_event,
+            t['crate'] + '/artifacts/' + t['runner'],
+            REPO_DIR,
+            corpus_vers,
+            bucket,
+            log_file,
+            True,
+            lambda path: report_artifact(path, branch, t),
+            artifact_deleted,
+        )
+
+        # Prepare the fuzz time metric
+        fuzz_time = FUZZ_TIME.labels(branch['name'], t['crate'], t['runner'], t['flags'])
+        last_time = time.monotonic()
+
+        # Finally, spin up the fuzzer itself
+        proc = subprocess.Popen(
+            ['cargo', 'fuzz', 'run', t['runner'], '--'] + t['flags'],
+            cwd = REPO_DIR / t['crate'],
+            start_new_session = True,
+            stdout = log_file,
+            stderr = subprocess.STDOUT,
+        )
+        fuzzer = {
+            'proc': proc,
+            'exit_event': exit_event,
+            'crate': t['crate'],
+            'runner': t['runner'],
+        }
+        FUZZERS.append(fuzzer)
+
+        # Wait for the fuzzer to complete (ie. either crash or requested to stop)
+        while proc.poll() == None and not exit_event.is_set():
+            time.sleep(0.5)
+            new_time = time.monotonic()
+            fuzz_time.inc(new_time - last_time)
+            last_time = new_time
```

Comment:
So the reason why I didn't use `os.wait` here was in order to be able to update the fuzzing time metric before the end of the process, otherwise the metric would basically stay at zero and after a day jump to one day of fuzzing, which wouldn't be nice.

Now it would probably be possible to have a single thread handling a list of fuzzer processes and then polling on each process by itself, without having one monitoring thread per process. My gut feeling had been that it'd make the code more straightforward to have one semantic thread per fuzzer process and then they each spin up/down their inotify threads and similar as they wish, which is the reason why I went with this.

That said, for further improvements (like avoiding to run two rsync's in parallel on the same corpus directory if the same fuzzer were to run twice on a single machine, which leads to errors though the corpus gets properly synced in the end) it'll probably be necessary to “unify” the fetching and pushing parts at least.

Do you feel like this should be done now, or as a follow-up iteration?

Reply:
My thinking is that having a single thread which monitors the fuzzers and also handles refreshes would be simpler because there’d be much less need for communication between threads. You would just have an os.wait loop and after AUTO_REFRESH_INTERVAL_SECS restart the fuzzers.

Reply:
By the way, at any event, `proc.wait(timeout=1)` would get rid of `time.sleep()`.